### ISSUE TITLE
About DCH evaluation metrics

### ISSUE THREAD

**Issue by liut0012** [thulab/DeepHash#14]
Hi,  
train_val_script.py first generates a complete list of rec, prec, mAP for different values of radius, and then a final set of metrics: 
i2i_after_sign 
i2i_prec_radius_2 
i2i_by_feature 
i2i_map_radius_2 
i2i_recall_radius_2 

As far as I understand "i2i_map_radius_2", "i2i_recall_radius_2" and "i2i_prec_radius_2" is the rec, prec, and mAP with the initial pruning Hamming radius setting to 2. Therefore, the values should be the same as the intermediate results generated during validation. 

In other words, the function "get_precision_recall_by_Hamming_Radius_All" and the "get_precision_recall_by_Hamming_Radius" should generate the same mAP for the same value of radius. 

However, this is not the case in my experiment. Could you help check these two codes? Or help me clarify if I have any wrong understanding?

**Comment by bl0** [comment]
Hi, can you send your log to my email(liubinthss@gmail.com)?

**Comment by bl0** [comment]
Actually the implementation of `get_precision_recall_by_Hamming_Radius_All` and `get_precision_recall_by_Hamming_Radius ` are slightly different. The implementation of `get_precision_recall_by_Hamming_Radius` is totally right while that of `get_precision_recall_by_Hamming_Radius_All` maybe not strictly accurate. but it doesn't have an obvious impact on our result since it's unlikely for our method to find 0 relevant code in the database similar to a query.

For your case, the precision is very low, then there are many 0 relevant retrievals, so the mAP results are different significantly.  If you train the model a bit iteration, the mAP results of the above two methods will be close.

Thanks very much!

**Comment by bl0** [comment]
I have fixed this.

Thanks again.

**Comment by liut0012** [comment]
Hi, you are right that there are many 0 relevant retrievals in my experimental results, because the class size of my data is quite small. 

So do you mean the discrepancy between the results given by the two functions is caused by how they handle the case when "all_num ==0"? I think this might not be the reason, because they are essentially handling it in the same way: 
* `get_precision_recall_by_Hamming_Radius` appends zero values 
https://github.com/thulab/DeepHash/blob/990d1336673636b0529dca7acd70d703b4b01231/DeepHash/evaluation/__init__.py#L135-L138
*  `get_precision_recall_by_Hamming_Radius_all` doesn't update the value, so leaving it to the default value, which is also zero.  https://github.com/thulab/DeepHash/blob/990d1336673636b0529dca7acd70d703b4b01231/DeepHash/evaluation/__init__.py#L151-L153

Eventually, I hope to modify the code so that the evaluation metrics given by the two functions are the same for any data. And I am still not sure how to do it...

**Comment by bl0** [comment]
Hi there, I found the issues. The calculation of `get_precision_recall_by_Hamming_Radius` is based on radius, which means the data points in the same distance contribute equally to mAP. while that of `get_precision_recall_by_Hamming_Radius_All` is based on the data point, which means even two data points have the same distance, they will have different weight to the final mAP because of the order. 

For the bellowing simple example, `get_precision_recall_by_Hamming_Radius` will treat the first 3 data points equally, while `get_precision_recall_by_Hamming_Radius_All` not.

|distance|0|0|0|1|
|--|--|--|--|--|
|correct|0|1|0|0|

So you can modify the `get_precision_recall_by_Hamming_Radius` to based on the radius.

Actually `get_mAPs` has the same issue, but it must return the `mAP` value based on `Rs`, so I don't have a good idea to deal with the conflict.

**Comment by liut0012** [comment]
My application prefers the way 'get_precision_recall_by_Hamming_Radius' handles it, i.e., treating the data points equally. I think I will just ignore the results given by 'get_precision_recall_by_Hamming_Radius_All'. Thanks a lot!

**Comment by liut0012** [comment]
Hi, in DCH paper, the evaluation method is described as: 
> We follow standard evaluation methods for Hamming space retrieval [9], which consists of two consecutive steps: (1) Pruning, to return data points within Hamming radius 2 for each query using hash table lookups; (2) Scanning, to re-rank the returned data points in ascending order of their distances to each query using continuous codes.

May I know which function is used to generate the mAP in this setting? It seems neither of the `get_precision_recall_by_Hamming_Radius_All` and 'get_precision_recall_by_Hamming_Radius' conducts the scanning (/reranking) step? 'get_mAPs_after_sign_with_feature_rerank' seem doing the job, but it does not accept the Hamming Radius as a parameter...

**Comment by bl0** [comment]
Yes, `get_mAPs_after_sign_with_feature_rerank` only calculate the reranked mAP with radius=2.

**Comment by liut0012** [comment]
Hi, did you use `get_mAPs_after_sign_with_feature_rerank` to generate the mAP reported in the paper? 

ps. I think you have two parameters to limit the number of retrieved items: 1) Rs, which is the number of retrieved items; 2) radius, which is the maximum distance of the retrieved items from the query. 
I think `get_mAPs_after_sign_with_feature_rerank`  accepts Rs, rather than radius (which is the one used in the paper). So I don't think `get_mAPs_after_sign_with_feature_rerank` is the one used to report results in the CVPR DCH paper. That's why I am asking... I want to know out of the five functions in the evaluation module which one should I use to...

Thanks for your prompt reply!

**Comment by bl0** [comment]
Yes
Best Regards,

Bin Liu
Address: Room 11-419, East Main Building,
               Tsinghua University, Beijing, 100084
                P.R. CHINA
E-mail: liubinthss@gmail.com

**Comment by bl0** [comment]
You can reference to:
https://github.com/thulab/DeepHash/blob/8cc84521cc2fb41a86c94e8db3ad8c3f9d9a2ed6/DeepHash/model/dch/dch.py#L251-L259
